### PR TITLE
Update dispatch stub to make SDPA routing cleaner

### PR DESCRIPTION
**Pull request by drisspg**
# Summary

Adds a public method to dispatchstub to check if a fn has been registered for a device. We use this new function to clean up the dispatching logic for SDPA, as well as make the private use dispatching simpler:
#126392

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
Very cool!

I guess this is just missing an implementation for the meta device to point it to the math implementation and it's ready to rock?

**Review by ezyang — APPROVED** [review]
okey dokey

**Comment by drisspg** [comment]
Hmmm, yeah so from the tests there are two devices that this function are 'get_call_ptr' is being called with that dont actually have an allocated fn_ptr for.. 'meta' and 'xla'

Options;
1. Add the ability to 'register_dispatch' for these functions'
2. Much less code (maybe less intuitive) if 'get_call_ptr' is called with a non supported device return a null_ptr


I think I am leaning towards two..

**Comment by drisspg** [comment]
2 works, but feels kinda sketchy/ bc breaking... hmm

I also see that this dispatch stub pretty heavily used so not sure if its faux pas to add extra members - more devices here

File: aten/src/ATen/native/DispatchStub.h [hunk 1]
```diff
@@ -174,6 +187,25 @@ struct DispatchStub<rT (*)(Args...), T> {
     impl.privateuse1_dispatch_ptr = reinterpret_cast<void*>(fn_ptr);
   }
 
+  // Returns true if the dispatcher has a kernel registered for this device
+  // type.
+  bool is_device_supported(const c10::DeviceType device_type) {
+    constexpr auto supported_devices = c10::array_of<c10::DeviceType>(
```
Comment (drisspg) [inline aten/src/ATen/native/DispatchStub.h:193]: https://quick-bench.com/q/4Fv3_mcsiNyQKvHedYYwZ_F1fvc

**Comment by drisspg** [comment]
Kind of audibled and defined the list of allowed devices types instead of refactoring get_code_ptr to update its default branch. Wanted this to be a lil more DRY but since users will have to manually add new device_ptr to the struct I think its okay as is

**Comment by drisspg** [comment]
ughhh, these damnn debug checks, I cant think of a nice way to implement this, I want to use template specialization .. let me sleep on this but this one is annoying without larger refactors

**Review by albanD — APPROVED** [review]
Nice!

**Pull request closed without merging**